pages/posts/validating-openapi-and-json-schema.md

```diff
@@ -0,0 +1,299 @@
+  console.log(result.valid);
+}());
+```
+
+#### With OpenAPI Schema dialect schema validation
+```javascript
+const OasSchema = require("@hyperjump/oas-schema-validator");
+const example = require("./example.openapi.json");
+
+(async function () {
+  const schema = await OasSchema.get("https://spec.openapis.org/oas/3.1/schema-base");
+  const validate = await OasSchema.validate(schema);
+
+  const result = validate(example);
+  console.log(result.valid);
+}());
+```
+
+### How does it work?
+
+To get an idea about how this works, let's take a look at a few selections from
+of the OpenAPI 3.1 schemas.
+
+This is where the Schema Object is defined. The `$dynamicAnchor` declares this
+sub-schema to be something that can be effectively overridden by another schema.
+If it's not overridden, the default behavior is to validate that the value is an
+object or a boolean. No other validation is performed on the schema.
+
+```yaml
+$defs:
+  schema:
+    $dynamicAnchor: meta
+    type:
+      - object
+      - boolean
+```
+
+When something in this schema wants to reference the Schema Object, instead of
+referencing `#/$defs/schema` like normal, it uses a dynamic reference to the
+"meta" dynamic anchor set in the previous selection. Now instead of always
+resolving to `#/$defs/schema`, another schema can potentially override where it
+resolves to.
+
+```yaml
+$defs:
+  components:
+    type: object
+    properties:
+      schemas:
+        type: object
+        additionalProperties:
+          $dynamicRef: '#meta'
+```
+
+## Validating Schema Objects against the default dialect
+
+With those vague building blocks in mind let's derive a schema that "extends"
+the abstract schema to create a schema that validates Schema Objects using the
+default dialect meta-schema.
+
+The first step is to include the abstract schema.
+```yaml
+$schema: 'https://json-schema.org/draft/2020-12/schema'
+
+$ref: 'https://spec.openapis.org/oas/3.1/schema/2021-09-28'
+```
+
+Then we need to add a `$dynamicAnchor` that matches the one in the abstract
+schema to override where dynamic references to "meta" will resolve to. From
+there we can reference the meta schema for the default dialect.
+
+```yaml
+$schema: 'https://json-schema.org/draft/2020-12/schema'
+
+$ref: 'https://spec.openapis.org/oas/3.1/schema/2021-09-28'
+
+$defs:
+  schema:
+    $dynamicAnchor: meta
+    $ref: 'https://spec.openapis.org/oas/3.1/dialect/base'
+```
+
+That's enough to get the Schema Object validation we were after, but there are a
+few loose ends we'll want to tie up as well. The `jsonSchemaDialect` field in
+the OpenAPI 3.1 document can be used to change the dialect used. Since this
+schema only supports the default dialect, we want to restrict people from
+changing that to something else. If they need to change it, they'll need a
+different schema to validate against. We also don't want people using the
+`$schema` keyword to change the dialect of individual schemas.
```

Comment:
Is this intending to imply that it's prohibited to switch to a different metaschema by using the `$schema` keyword in an individual schema? I don't think that's correct -- [the docs for jsonSchemaDialect](https://spec.openapis.org/oas/v3.1.0#openapi-object) just say that this is the *default* value for the metaschema, not the only allowable value.

Reply:
(and indeed a subsequent example in this article goes on to show how to use a different metaschema for one particular schema.)

Reply:
I spent a little time thinking about this, and the intended implication is the context of "in this situation...". A hypothetical "we" which has imposed a specific governance rule for the work.

Reply:
ah okay, fair enough.